Sources/ApolloCodegenLib/Templates/TemplateRenderer.swift

```diff
@@ -0,0 +1,104 @@
+import ApolloUtils
+
+// MARK: TemplateRenderer
+
+/// Defines the file target of the template.
+enum TemplateTarget {
+  /// Used in schema types files; enum, input object, union, etc.
+  case schemaFile
+  /// Used in operation files; query, mutation, fragment, etc.
+  case operationFile
+  /// Used in files that define a module; Swift Package Manager, etc.
+  case moduleFile
+}
+
+/// A protocol to handle the rendering of a file template based on the target file type and
+/// codegen configuration.
+///
+/// All templates that output to a file should conform to this protocol, this does not include
+/// templates that are used by others such as `HeaderCommentTemplate` or `ImportStatementTemplate`.
+protocol TemplateRenderer {
+  /// File target of the template.
+  var target: TemplateTarget { get }
+  /// The swift code format.
+  var template: TemplateString { get }
+
+  /// Renders the template converting all input values and generating a final String representation
+  /// of the template.
+  ///
+  /// - Parameter config: Shared codegen configuration.
+  /// - Returns: Swift code derived from the template format.
+  func render(forConfig config: ReferenceWrapped<ApolloCodegenConfiguration>) -> String
```

Comment:
I don't think this should be a function in the protocol and instead just have the extension implement it. Having it as part of the protocol might lead someone to creating a new template to implement it, which would potentially cause problems for the new template.